{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gapfilling with Sco Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/te/Library/CloudStorage/OneDrive-Personal/Biosustain/strep/Code/streptsd/notebooks/gapfilling/1.gapfill_w_sco\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'' is not a valid SBML 'SId'.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2024-11-13\n"
     ]
    }
   ],
   "source": [
    "import cobra\n",
    "from cobra.core import Metabolite, Reaction\n",
    "from cobra.io import read_sbml_model, load_json_model\n",
    "from cobra.flux_analysis import gapfill\n",
    "import os, re\n",
    "import pandas as pd\n",
    "\n",
    "print(os.getcwd())\n",
    "\n",
    "ext_dir = '/../../../data/external'\n",
    "phenomics = '/../../phenomics'\n",
    "\n",
    "# Import model\n",
    "# model = read_sbml_model(f\"{os.getcwd()}/{ext_dir}/Salb-GEM.xml\")\n",
    "model = read_sbml_model(f\"{os.getcwd()}/{ext_dir}/Salb-GEM-Updated.xml\")\n",
    "\n",
    "# Mapping of the sco_model reaction model\n",
    "sco_model = read_sbml_model(f\"{os.getcwd()}/{ext_dir}/Sco-GEM.xml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Get all untestable and false negative compounds "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>activity</th>\n",
       "      <th>chemical</th>\n",
       "      <th>category</th>\n",
       "      <th>moa</th>\n",
       "      <th>bigg</th>\n",
       "      <th>exchange</th>\n",
       "      <th>model_simulation_0.05</th>\n",
       "      <th>model_simulation_0.1</th>\n",
       "      <th>model_simulation_0.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PM01_A01</td>\n",
       "      <td>False</td>\n",
       "      <td>Negative Control</td>\n",
       "      <td>carbon</td>\n",
       "      <td>C-Source, negative control</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PM01_A02</td>\n",
       "      <td>True</td>\n",
       "      <td>L-Arabinose</td>\n",
       "      <td>carbon</td>\n",
       "      <td>C-Source, carbohydrate</td>\n",
       "      <td>arab__L</td>\n",
       "      <td>EX_arab__L_e</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PM01_A03</td>\n",
       "      <td>True</td>\n",
       "      <td>N-Acetyl-D-Glucosamine</td>\n",
       "      <td>carbon</td>\n",
       "      <td>C-Source, carbohydrate</td>\n",
       "      <td>acgam</td>\n",
       "      <td>EX_acgam_e</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PM01_A04</td>\n",
       "      <td>False</td>\n",
       "      <td>D-Saccharic acid</td>\n",
       "      <td>carbon</td>\n",
       "      <td>C-Source, carboxylic acid</td>\n",
       "      <td>glcr</td>\n",
       "      <td>EX_glcr_e</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PM01_A05</td>\n",
       "      <td>True</td>\n",
       "      <td>Succinic acid</td>\n",
       "      <td>carbon</td>\n",
       "      <td>C-Source, carboxylic acid</td>\n",
       "      <td>succ</td>\n",
       "      <td>EX_succ_e</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>PM04_H08</td>\n",
       "      <td>False</td>\n",
       "      <td>p-Aminobenzene Sulfonic acid</td>\n",
       "      <td>phosphate &amp; sulphur</td>\n",
       "      <td>S-Source, organic</td>\n",
       "      <td>4abz</td>\n",
       "      <td>EX_4abz_e</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>PM04_H09</td>\n",
       "      <td>False</td>\n",
       "      <td>Butane Sulfonic acid</td>\n",
       "      <td>phosphate &amp; sulphur</td>\n",
       "      <td>S-Source, organic</td>\n",
       "      <td>butso3</td>\n",
       "      <td>EX_butso3_e</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>PM04_H10</td>\n",
       "      <td>False</td>\n",
       "      <td>2-Hydroxyethane Sulfonic acid</td>\n",
       "      <td>phosphate &amp; sulphur</td>\n",
       "      <td>S-Source, organic</td>\n",
       "      <td>isetac</td>\n",
       "      <td>EX_isetac_e</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>PM04_H11</td>\n",
       "      <td>False</td>\n",
       "      <td>Methane Sulfonic acid</td>\n",
       "      <td>phosphate &amp; sulphur</td>\n",
       "      <td>S-Source, organic</td>\n",
       "      <td>mso3</td>\n",
       "      <td>EX_mso3_e</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383</th>\n",
       "      <td>PM04_H12</td>\n",
       "      <td>False</td>\n",
       "      <td>Tetramethylene Sulfone</td>\n",
       "      <td>phosphate &amp; sulphur</td>\n",
       "      <td>S-Source, organic</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>384 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        index  activity                       chemical             category  \\\n",
       "0    PM01_A01     False               Negative Control               carbon   \n",
       "1    PM01_A02      True                    L-Arabinose               carbon   \n",
       "2    PM01_A03      True         N-Acetyl-D-Glucosamine               carbon   \n",
       "3    PM01_A04     False               D-Saccharic acid               carbon   \n",
       "4    PM01_A05      True                  Succinic acid               carbon   \n",
       "..        ...       ...                            ...                  ...   \n",
       "379  PM04_H08     False   p-Aminobenzene Sulfonic acid  phosphate & sulphur   \n",
       "380  PM04_H09     False           Butane Sulfonic acid  phosphate & sulphur   \n",
       "381  PM04_H10     False  2-Hydroxyethane Sulfonic acid  phosphate & sulphur   \n",
       "382  PM04_H11     False          Methane Sulfonic acid  phosphate & sulphur   \n",
       "383  PM04_H12     False         Tetramethylene Sulfone  phosphate & sulphur   \n",
       "\n",
       "                            moa     bigg      exchange  model_simulation_0.05  \\\n",
       "0    C-Source, negative control        #             #                   <NA>   \n",
       "1        C-Source, carbohydrate  arab__L  EX_arab__L_e                  False   \n",
       "2        C-Source, carbohydrate    acgam    EX_acgam_e                   <NA>   \n",
       "3     C-Source, carboxylic acid     glcr     EX_glcr_e                   <NA>   \n",
       "4     C-Source, carboxylic acid     succ     EX_succ_e                   True   \n",
       "..                          ...      ...           ...                    ...   \n",
       "379           S-Source, organic     4abz     EX_4abz_e                   <NA>   \n",
       "380           S-Source, organic   butso3   EX_butso3_e                   <NA>   \n",
       "381           S-Source, organic   isetac   EX_isetac_e                   <NA>   \n",
       "382           S-Source, organic     mso3     EX_mso3_e                   <NA>   \n",
       "383           S-Source, organic        #             #                   <NA>   \n",
       "\n",
       "    model_simulation_0.1 model_simulation_0.2  \n",
       "0                    NaN                  NaN  \n",
       "1                  False                False  \n",
       "2                    NaN                  NaN  \n",
       "3                    NaN                  NaN  \n",
       "4                   True                 True  \n",
       "..                   ...                  ...  \n",
       "379                  NaN                  NaN  \n",
       "380                  NaN                  NaN  \n",
       "381                  NaN                  NaN  \n",
       "382                  NaN                  NaN  \n",
       "383                  NaN                  NaN  \n",
       "\n",
       "[384 rows x 10 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import agreed biolog vs model data\n",
    "agreed_bio_data = pd.read_csv(f\"{os.getcwd()}{phenomics}/agreed_bio_data_Salb.csv\")\n",
    "agreed_bio_data['model_simulation_0.05'] = agreed_bio_data['model_simulation_0.05'].astype('boolean')\n",
    "\n",
    "# # Mapping between bigg and metax names\n",
    "# bigg_mnmx = pd.read_json(f\"{os.getcwd()}/{ext_dir}/metanetx.json\")\n",
    "\n",
    "# # Mapping files for bigg compound to all other names\n",
    "# name_map = pd.read_csv(f\"{os.getcwd()}/{ext_dir}/biggmodels_metabolites.txt\",sep='\\t')\n",
    "# name_map = name_map.drop(['bigg_id', 'model_list', 'old_bigg_ids'], axis=1)\n",
    "\n",
    "agreed_bio_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "False negative (actual true, predicted false):\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(f\"\\nFalse negative (actual true, predicted false):\\n\")\n",
    "agreed_bio_data_nan_drop = agreed_bio_data[agreed_bio_data['model_simulation_0.05'].notna()]\n",
    "false_negatives = agreed_bio_data_nan_drop[\n",
    "    agreed_bio_data_nan_drop.xs(\"activity\", axis=1)\n",
    "    & ~(agreed_bio_data_nan_drop.xs(\"model_simulation_0.05\", axis=1))\n",
    "    ][[\"bigg\",\"chemical\", \"moa\", \"exchange\", \"model_simulation_0.05\"]]\n",
    "\n",
    "false_negatives\n",
    "\n",
    "false_negatives.to_csv(os.getcwd() + \"/\" + 'Salb_false_negative.csv', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Try Gapfilling Function on false negatives"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def basic_gapfill(bigg, model, reference, objective='growth', iter=1):\n",
    "    \"\"\"\n",
    "    gapfill a model using reference model and cplex solver.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    bigg: String\n",
    "    model: Model\n",
    "    reference: Model\n",
    "    \"\"\"\n",
    "    print(f\"\\nGapfilling {bigg} with: \")\n",
    "    with reference:\n",
    "        reference.solver = 'cplex'\n",
    "        model.solver = 'cplex'\n",
    "        model.objective = objective\n",
    "        solution = gapfill(model, reference, demand_reactions=False, iterations=iter)\n",
    "        print(solution)\n",
    "    return solution \n",
    "\n",
    "def gapfill_medium(model, reference, bigg, type, exchange):\n",
    "    \"\"\"\n",
    "    This function does gapfilling based on a chemical in the certain biolog medium and add an exchange reaction if needed\"\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    bigg (str) \n",
    "    category (str)\n",
    "    exchange (str)\n",
    "    \"\"\"\n",
    "\n",
    "    # when model has the exchange reaction, no need to add a reaction before gapfilling.  \n",
    "    try:\n",
    "        with model:\n",
    "            medium = model.medium\n",
    "            medium[exchange] = 0.8\n",
    "            if type.startswith(\"C\"):\n",
    "                medium[\"EX_glc__D_e\"] = 0\n",
    "\n",
    "            elif type.startswith(\"N\"):\n",
    "                medium[\"EX_nh4_e\"] = 0\n",
    "\n",
    "            elif type.startswith(\"P\"):\n",
    "                medium[\"EX_pi_e\"] = 0\n",
    "            \n",
    "            elif type.startswith(\"S\"):\n",
    "                medium[\"EX_so4_e\"] = 0\n",
    "\n",
    "            model.medium = medium\n",
    "            model.reactions.EX_co2_e.bounds= (0, 1000)\n",
    "\n",
    "            return basic_gapfill(bigg, model, reference, 'growth', 4)\n",
    "    except:\n",
    "        print(f\"Gapfilling of the compound {bigg} - {type} - {exchange} is failed\\n\")\n",
    "        return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmplopuhu5q.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "\n",
      "Gapfilling arab__L with: \n",
      "[[<Reaction ARAI at 0x7f7812414c40>, <Reaction DTDP26DGR at 0x7f77e495f820>], [<Reaction ARAI at 0x7f7812414c40>, <Reaction DTDP26DGR at 0x7f77e495f820>], [<Reaction ARAI at 0x7f7812414c40>, <Reaction DTDP26DGR at 0x7f77e495f820>], [<Reaction ARAI at 0x7f7812414c40>, <Reaction DTDP26DGR at 0x7f77e495f820>]]\n",
      "DTDP26DGR, ARAI\n",
      "\n",
      "Gapfilling man with: \n",
      "[[<Reaction MANpts at 0x7f77f48791f0>], [<Reaction HEX4 at 0x7f77f48437c0>], [<Reaction HEX4 at 0x7f77f48437c0>], [<Reaction MANpts at 0x7f77f48791f0>]]\n",
      "HEX4, MANpts\n",
      "\n",
      "Gapfilling lyx__L with: \n",
      "[[<Reaction DTDP26DGR at 0x7f7812d977f0>, <Reaction FBA4 at 0x7f77f56393a0>, <Reaction KHK2 at 0x7f7813182cd0>, <Reaction LYXI at 0x7f7812e4f250>], [<Reaction DTDP26DGR at 0x7f7812d977f0>, <Reaction FBA4 at 0x7f77f56393a0>, <Reaction KHK2 at 0x7f7813182cd0>, <Reaction LYXI at 0x7f7812e4f250>], [<Reaction DTDP26DGR at 0x7f7812d977f0>, <Reaction FBA4 at 0x7f77f56393a0>, <Reaction KHK2 at 0x7f7813182cd0>, <Reaction LYXI at 0x7f7812e4f250>], [<Reaction DTDP26DGR at 0x7f7812d977f0>, <Reaction FBA4 at 0x7f77f56393a0>, <Reaction KHK2 at 0x7f7813182cd0>, <Reaction LYXI at 0x7f7812e4f250>]]\n",
      "FBA4, LYXI, KHK2, DTDP26DGR\n",
      "\n",
      "Gapfilling 5dglcn with: \n",
      "[[<Reaction 5DGLCNR at 0x7f78140f0a90>, <Reaction DTDP26DGR at 0x7f779006b8b0>], [<Reaction 5DGLCNR at 0x7f78140f0a90>, <Reaction DTDP26DGR at 0x7f779006b8b0>], [<Reaction 5DGLCNR at 0x7f78140f0a90>, <Reaction DTDP26DGR at 0x7f779006b8b0>], [<Reaction 5DGLCNR at 0x7f78140f0a90>, <Reaction DTDP26DGR at 0x7f779006b8b0>]]\n",
      "5DGLCNR, DTDP26DGR\n",
      "\n",
      "Gapfilling met__L with: \n",
      "Gapfilling of the compound met__L - C-Source, amino acid - EX_met__L_e is failed\n",
      "\n",
      "\n",
      "Gapfilling phe__L with: \n",
      "Gapfilling of the compound phe__L - C-Source, amino acid - EX_phe__L_e is failed\n",
      "\n",
      "\n",
      "Gapfilling no2 with: \n",
      "[[<Reaction DTDP26DGR at 0x7f7806e9c550>, <Reaction NO2t2r at 0x7f77a197fa00>], [<Reaction DTDP26DGR at 0x7f7806e9c550>, <Reaction NO2t2r at 0x7f77a197fa00>], [<Reaction DTDP26DGR at 0x7f7806e9c550>, <Reaction NO2t2r at 0x7f77a197fa00>], [<Reaction DTDP26DGR at 0x7f7806e9c550>, <Reaction NO2t2r at 0x7f77a197fa00>]]\n",
      "NO2t2r, DTDP26DGR\n",
      "\n",
      "Gapfilling arg__L with: \n",
      "[[<Reaction CITRLt at 0x7f780758e130>, <Reaction EX_citr__L_e at 0x7f77923822e0>], [<Reaction CITRLt at 0x7f780758e130>, <Reaction EX_citr__L_e at 0x7f77923822e0>], [<Reaction CITRLt at 0x7f780758e130>, <Reaction EX_citr__L_e at 0x7f77923822e0>], [<Reaction CITRLt at 0x7f780758e130>, <Reaction EX_citr__L_e at 0x7f77923822e0>]]\n",
      "CITRLt, EX_citr__L_e\n",
      "\n",
      "Gapfilling cys__L with: \n",
      "[[<Reaction H2St1 at 0x7f77931665e0>, <Reaction EX_h2s_e at 0x7f77aa2f02e0>], [<Reaction H2St1 at 0x7f77931665e0>, <Reaction EX_h2s_e at 0x7f77aa2f02e0>], [<Reaction H2St1 at 0x7f77931665e0>, <Reaction EX_h2s_e at 0x7f77aa2f02e0>], [<Reaction H2St1 at 0x7f77931665e0>, <Reaction EX_h2s_e at 0x7f77aa2f02e0>]]\n",
      "EX_h2s_e, H2St1\n",
      "\n",
      "Gapfilling lys__L with: \n",
      "[[<Reaction APTNAT at 0x7f77a42a5d90>, <Reaction DAPAT at 0x7f7791996820>, <Reaction PPRDNDH at 0x7f7815e43940>], [<Reaction APTNAT at 0x7f77a42a5d90>, <Reaction DAPAT at 0x7f7791996820>, <Reaction PPRDNDH at 0x7f7815e43940>], [<Reaction APTNAT at 0x7f77a42a5d90>, <Reaction DAPAT at 0x7f7791996820>, <Reaction PPRDNDH at 0x7f7815e43940>], [<Reaction APTNAT at 0x7f77a42a5d90>, <Reaction DAPAT at 0x7f7791996820>, <Reaction PPRDNDH at 0x7f7815e43940>]]\n",
      "APTNAT, DAPAT, PPRDNDH\n",
      "\n",
      "Gapfilling met__L with: \n",
      "[[<Reaction ACONIs at 0x7f77902da6a0>, <Reaction DTDP26DGR at 0x7f77adb75ac0>, <Reaction H2St1 at 0x7f77945772b0>, <Reaction EX_h2s_e at 0x7f77adb49f70>], [<Reaction ACONIs at 0x7f77902da6a0>, <Reaction DTDP26DGR at 0x7f77adb75ac0>, <Reaction H2St1 at 0x7f77945772b0>, <Reaction EX_h2s_e at 0x7f77adb49f70>], [<Reaction ACONIs at 0x7f77902da6a0>, <Reaction DTDP26DGR at 0x7f77adb75ac0>, <Reaction H2St1 at 0x7f77945772b0>, <Reaction EX_h2s_e at 0x7f77adb49f70>], [<Reaction 2METBOD at 0x7f7816485d00>, <Reaction DTDP26DGR at 0x7f77adb75ac0>, <Reaction GERDPM at 0x7f77adadd820>, <Reaction H2St1 at 0x7f77945772b0>, <Reaction DM_2mborn_c at 0x7f77adb2f1f0>, <Reaction EX_h2s_e at 0x7f77adb49f70>]]\n",
      "GERDPM, ACONIs, H2St1, DTDP26DGR, 2METBOD, EX_h2s_e, DM_2mborn_c\n",
      "\n",
      "Gapfilling phe__L with: \n",
      "[[<Reaction PPYRDC at 0x7f7796d32d90>], [<Reaction PHYCBOXL at 0x7f7796d32340>], [<Reaction PHYCBOXL at 0x7f7796d32340>], [<Reaction PPYRDC at 0x7f7796d32d90>]]\n",
      "PPYRDC, PHYCBOXL\n",
      "\n",
      "Gapfilling tyr__L with: \n",
      "[[<Reaction MACACI at 0x7f77aeb18850>], [<Reaction MACACI at 0x7f77aeb18850>], [<Reaction MACACI at 0x7f77aeb18850>], [<Reaction MACACI at 0x7f77aeb18850>]]\n",
      "MACACI\n",
      "\n",
      "Gapfilling hom__L with: \n",
      "Gapfilling of the compound hom__L - N-Source, amino acid - EX_hom__L_e is failed\n",
      "\n",
      "\n",
      "Gapfilling etha with: \n",
      "[[<Reaction ETHAAL at 0x7f77999720d0>], [<Reaction ETHAAL at 0x7f77999720d0>], [<Reaction ETHAAL at 0x7f77999720d0>], [<Reaction ETHAAL at 0x7f77999720d0>]]\n",
      "ETHAAL\n",
      "\n",
      "Gapfilling thym with: \n",
      "Gapfilling of the compound thym - N-Source, other - EX_thym_e is failed\n",
      "\n",
      "\n",
      "Gapfilling thymd with: \n",
      "Gapfilling of the compound thymd - N-Source, other - EX_thymd_e is failed\n",
      "\n",
      "\n",
      "Gapfilling ura with: \n",
      "Gapfilling of the compound ura - N-Source, other - EX_ura_e is failed\n",
      "\n",
      "\n",
      "Gapfilling uri with: \n",
      "Gapfilling of the compound uri - N-Source, other - EX_uri_e is failed\n",
      "\n",
      "\n",
      "Gapfilling glyc3p with: \n",
      "Gapfilling of the compound glyc3p - P-Source, organic - EX_glyc3p_e is failed\n",
      "\n"
     ]
    }
   ],
   "source": [
    "_model = model.copy()\n",
    "with open('gapfill_reactions.txt', 'w') as file:\n",
    "    for ind in false_negatives.index:\n",
    "        file.write(f\"Gapfilling {false_negatives['bigg'][ind]}, {false_negatives['moa'][ind]} with:\\n\")\n",
    "        solution = gapfill_medium(_model, sco_model, false_negatives['bigg'][ind], false_negatives['moa'][ind], false_negatives['exchange'][ind])\n",
    "        if solution:\n",
    "            unique_solution = set([reaction for reaction_set in solution for reaction in reaction_set])\n",
    "            reactions_str = ', '.join(str(reaction.id) for reaction in unique_solution)\n",
    "            print(reactions_str)\n",
    "            file.write(reactions_str + '\\n\\n')\n",
    "        else:\n",
    "            file.write('\\n\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'DTDP26DGR', 'ACONIs', 'GERDPM', 'PPRDNDH', 'DM_2mborn_c', 'CITRLt', 'HEX4', 'DAPAT', 'EX_h2s_e', 'FBA4', 'APTNAT', '5DGLCNR', 'EX_citr__L_e', 'MACACI', '2METBOD', 'ETHAAL', 'ARAI', 'PHYCBOXL', 'H2St1', 'MANpts', 'PPYRDC', 'LYXI', 'KHK2', 'NO2t2r'}\n"
     ]
    }
   ],
   "source": [
    "# Initialize an empty list to store the extracted lists\n",
    "gapfill_lists = []\n",
    "\n",
    "# Open and read the file\n",
    "with open('gapfill_reactions.txt', 'r') as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "# Process every second line and split it into a list\n",
    "for i in range(1, len(lines), 3):\n",
    "    reaction_list = [item.strip() for item in lines[i].strip().split(',')]\n",
    "    gapfill_lists.extend(reaction_list)\n",
    "\n",
    "merged_lists = set(gapfill_lists)\n",
    "merged_lists.discard('')\n",
    "\n",
    "# Print the extracted lists\n",
    "print(merged_lists)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Get a short list by testing new reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpz_1geyxz.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------\n",
      "Test with DTDP26DGR \n",
      "Solution: <Solution 0.156 at 0x7f7764778e50>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpnctgqbsk.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with ACONIs \n",
      "Solution: <Solution 0.052 at 0x7f77e3858bb0>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpfhsm9w3i.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with GERDPM \n",
      "Solution: <Solution 0.052 at 0x7f77830a4160>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpruxexr31.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with PPRDNDH \n",
      "Solution: <Solution 0.052 at 0x7f779d71a5e0>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmp7e6w1a3m.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with DM_2mborn_c \n",
      "Solution: <Solution 0.052 at 0x7f779d9017c0>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpd4kz7rm2.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with CITRLt \n",
      "Solution: <Solution 0.052 at 0x7f7765bd0370>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmp4hop7ult.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with HEX4 \n",
      "Solution: <Solution 0.052 at 0x7f779df3d700>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpy9up1im3.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with DAPAT \n",
      "Solution: <Solution 0.052 at 0x7f7777c46a00>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpab9pvf1y.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with EX_h2s_e \n",
      "Solution: <Solution 0.052 at 0x7f775427bbe0>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpqn4t5lt3.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with FBA4 \n",
      "Solution: <Solution 0.052 at 0x7f7766c54f10>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmp_40r6v5f.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with APTNAT \n",
      "Solution: <Solution 0.052 at 0x7f77657ce9d0>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmp37ohdhux.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with 5DGLCNR \n",
      "Solution: <Solution 0.052 at 0x7f7766f5dcd0>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmplm3ttbg_.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with EX_citr__L_e \n",
      "Solution: <Solution 0.052 at 0x7f7755e73e20>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmp_48c0r0k.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with MACACI \n",
      "Solution: <Solution 0.052 at 0x7f779f8af5b0>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpgjspsy27.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with 2METBOD \n",
      "Solution: <Solution 0.052 at 0x7f773066a3a0>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpc6nkslca.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with ETHAAL \n",
      "Solution: <Solution 0.052 at 0x7f7742ff9760>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpjqmkgtp_.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with ARAI \n",
      "Solution: <Solution 0.052 at 0x7f7743a4b850>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpoyagd2z3.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with PHYCBOXL \n",
      "Solution: <Solution 0.052 at 0x7f7730e629d0>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpd3ic5ja6.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with H2St1 \n",
      "Solution: <Solution 0.052 at 0x7f77113aabb0>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpw5i82bsj.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with MANpts \n",
      "Solution: <Solution 0.052 at 0x7f7711744a60>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmp8zr3934g.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with PPYRDC \n",
      "Solution: <Solution 0.052 at 0x7f77548a9820>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmp2sepu06d.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with LYXI \n",
      "Solution: <Solution 0.052 at 0x7f7744f85bb0>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmph7fxor81.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with KHK2 \n",
      "Solution: <Solution 0.052 at 0x7f773304ca30>\n",
      "--------\n",
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmplqkgtive.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1874 rows, 4466 columns, 19614 nonzeros\n",
      "------\n",
      "Test with NO2t2r \n",
      "Solution: <Solution 0.052 at 0x7f7712702430>\n",
      "--------\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['DTDP26DGR',\n",
       " 'ACONIs',\n",
       " 'GERDPM',\n",
       " 'PPRDNDH',\n",
       " 'DM_2mborn_c',\n",
       " 'CITRLt',\n",
       " 'HEX4',\n",
       " 'DAPAT',\n",
       " 'EX_h2s_e',\n",
       " 'FBA4',\n",
       " 'APTNAT',\n",
       " '5DGLCNR',\n",
       " 'EX_citr__L_e',\n",
       " 'MACACI',\n",
       " '2METBOD',\n",
       " 'ETHAAL',\n",
       " 'ARAI',\n",
       " 'PHYCBOXL',\n",
       " 'H2St1',\n",
       " 'MANpts',\n",
       " 'PPYRDC',\n",
       " 'LYXI',\n",
       " 'KHK2',\n",
       " 'NO2t2r']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "short_list = []\n",
    "for reaction_id in merged_lists:\n",
    "    # Get the reaction from the reference model\n",
    "    reaction = sco_model.reactions.get_by_id(reaction_id)\n",
    "    model_temp = model.copy()\n",
    "    \n",
    "    with model_temp:\n",
    "        # Add a copy of the reaction to your model\n",
    "        model_temp.add_reactions([reaction.copy()])\n",
    "        model_temp.solver ='cplex'\n",
    "        model_temp.reactions.EX_co2_e.bounds= (0, 1000)\n",
    "\n",
    "        solution = model_temp.optimize()\n",
    "        print (f\"------\\nTest with {reaction_id} \\nSolution: {solution}\\n--------\")\n",
    "\n",
    "        if solution.objective_value < 1:\n",
    "            short_list.append(reaction_id)\n",
    "\n",
    "short_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests, re\n",
    "\n",
    "def check_annotation_uniprot(reaction, taxonomy_id):\n",
    "    ec_numbers = reaction.annotation.get('ec-code', [])\n",
    "    result = []\n",
    "    for ec in ec_numbers:\n",
    "        # Ensure the EC number has only digits and dots\n",
    "        if re.match(r'^\\d+(\\.\\d+)+$', ec):\n",
    "            print(ec)\n",
    "            query = f'(taxonomy_id:{taxonomy_id}) AND (ec:{ec})'\n",
    "            url = 'https://rest.uniprot.org/uniprotkb/search?'\n",
    "            params = {\n",
    "                'query': query,\n",
    "                'format': 'list',\n",
    "            }\n",
    "            response = requests.get(url, params=params)\n",
    "            if response.text != '':\n",
    "                result.append(response.text)\n",
    "    print(result)\n",
    "    return result\n",
    "\n",
    "def check_transport_reaction(reaction):\n",
    "    compartments = {metabolite.compartment for metabolite in reaction.metabolites}\n",
    "    is_transport = len(compartments) > 1\n",
    "    \n",
    "    print(f\"transport reaction: {is_transport}\")\n",
    "   \n",
    "    return is_transport"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------\n",
      "HEX4: atp_c + man_c --> adp_c + h_c + man6p_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "FBA4: xu1p__D_c <=> dhap_c + gcald_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "PPRDNDH: 2.0 h2o_c + nad_c + pprdn_c --> 5aptn_c + h_c + nadh_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "2METBOD: 2mygydpp_c + h2o_c --> 2mborn_c + ppi_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "NO2t2r: h_e + no2_e <=> h_c + no2_c\n",
      "transport reaction: True\n",
      "--------\n",
      "5DGLCNR: 5dglcn_c + h_c + nadph_c <=> glcn_c + nadp_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "GERDPM: amet_c + grdp_c --> 2mygydpp_c + ahcys_c + h_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "EX_h2s_e: h2s_e --> \n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "MACACI: 4mlacac_c <=> 4fumacac_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "H2St1: h2s_e <=> h2s_c\n",
      "transport reaction: True\n",
      "--------\n",
      "ACONIs: acon_T_c <=> acon_C_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "APTNAT: 5aptn_c + akg_c <=> glu__L_c + oxptn_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "MANpts: man_e + pep_c --> man6p_c + pyr_c\n",
      "transport reaction: True\n",
      "--------\n",
      "EX_citr__L_e: citr__L_e --> \n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "LYXI: lyx__L_c <=> xylu__L_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "DM_2mborn_c: 2mborn_c --> \n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "CITRLt: citr__L_c <=> citr__L_e\n",
      "transport reaction: True\n",
      "--------\n",
      "KHK2: atp_c + xylu__L_c --> adp_c + xu1p__D_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "DAPAT: 15dap_c + akg_c --> glu__L_c + h2o_c + h_c + pprdn_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "DTDP26DGR: dtdp26ddgh2e2u_c + h_c --> dtdp34dh26doglc_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "ARAI: arab__L_c <=> rbl__L_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "PPYRDC: h_c + phpyr_c --> co2_c + pacald_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "ETHAAL: etha_c --> acald_c + nh4_c\n",
      "transport reaction: False\n",
      "[]\n",
      "--------\n",
      "PHYCBOXL: h_c + phe__L_c --> co2_c + peamn_c\n",
      "transport reaction: False\n",
      "[]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fact checking using different methods, if a reaction is not found in uniprot, then it's likely the reaction doesn't exist\n",
    "taxonomy_id = 1886\n",
    "\n",
    "final_list = []\n",
    "for r in short_list:\n",
    "    reaction = sco_model.reactions.get_by_id(r)\n",
    "    print(f'--------\\n{reaction}')\n",
    "    if check_transport_reaction(reaction) == False:\n",
    "        if reaction.annotation:\n",
    "            if check_annotation_uniprot(reaction, taxonomy_id):\n",
    "                final_list.append(r)\n",
    "\n",
    "final_list  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test added reactions\n",
    "for reaction_id in short_list:\n",
    "    # Get the reaction from the sco_model model\n",
    "    reaction = sco_model.reactions.get_by_id(reaction_id)\n",
    "    \n",
    "    # Add a copy of the reaction to your model\n",
    "    model.add_reactions([reaction.copy()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Optimize arab__L, C-Source, carbohydrate with: \n",
      "{'EX_arab__L_e': 0.8, 'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_nh4_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.13031365898552\n",
      "\n",
      " Optimize man, C-Source, carbohydrate with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_man_e': 0.8, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_nh4_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.15637639078262328\n",
      "\n",
      " Optimize lyx__L, C-Source, carbohydrate with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_lyx__L_e': 0.8, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_nh4_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.13031365898551833\n",
      "\n",
      " Optimize 5dglcn, C-Source, carboxylic acid with: \n",
      "{'EX_5dglcn_e': 0.8, 'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_nh4_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.1303136589855191\n",
      "\n",
      " Optimize met__L, C-Source, amino acid with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_met__L_e': 0.8, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_nh4_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.0\n",
      "\n",
      " Optimize phe__L, C-Source, amino acid with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_nh4_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_phe__L_e': 0.8, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.01321865679499643\n",
      "\n",
      " Optimize no2, N-Source, inorganic with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_glc__D_e': 0.8, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_no2_e': 0.8, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.1203798349189642\n",
      "\n",
      " Optimize arg__L, N-Source, amino acid with: \n",
      "{'EX_arg__L_e': 0.8, 'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_glc__D_e': 0.8, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.1406645803809539\n",
      "\n",
      " Optimize cys__L, N-Source, amino acid with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_cys__L_e': 0.8, 'EX_fe2_e': 1000.0, 'EX_glc__D_e': 0.8, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.1307386474621115\n",
      "\n",
      " Optimize lys__L, N-Source, amino acid with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_glc__D_e': 0.8, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_lys__L_e': 0.8, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.26147729492420735\n",
      "\n",
      " Optimize met__L, N-Source, amino acid with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_glc__D_e': 0.8, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_met__L_e': 0.8, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.13073864746210748\n",
      "\n",
      " Optimize phe__L, N-Source, amino acid with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_glc__D_e': 0.8, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_phe__L_e': 0.8, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.13073864746209626\n",
      "\n",
      " Optimize tyr__L, N-Source, amino acid with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_glc__D_e': 0.8, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_tyr__L_e': 0.8, 'EX_zn2_e': 1000.0}\n",
      "0.13073864746210045\n",
      "\n",
      " Optimize hom__L, N-Source, amino acid with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_glc__D_e': 0.8, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_hom__L_e': 0.8, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.0\n",
      "\n",
      " Optimize etha, N-Source, other with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_etha_e': 0.8, 'EX_fe2_e': 1000.0, 'EX_glc__D_e': 0.8, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.13073864746210642\n",
      "\n",
      " Optimize thym, N-Source, other with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_glc__D_e': 0.8, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_thym_e': 0.8, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.0\n",
      "\n",
      " Optimize thymd, N-Source, other with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_glc__D_e': 0.8, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_thymd_e': 0.8, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.0\n",
      "\n",
      " Optimize ura, N-Source, other with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_glc__D_e': 0.8, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_ura_e': 0.8, 'EX_zn2_e': 1000.0}\n",
      "0.0\n",
      "\n",
      " Optimize uri, N-Source, other with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_glc__D_e': 0.8, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_uri_e': 0.8, 'EX_zn2_e': 1000.0}\n",
      "0.0\n",
      "\n",
      " Optimize glyc3p, P-Source, organic with: \n",
      "{'EX_ca2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_glc__D_e': 0.8, 'EX_glyc3p_e': 0.8, 'EX_h2o_e': 1000.0, 'EX_h_e': 1000.0, 'EX_k_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_nh4_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_zn2_e': 1000.0}\n",
      "0.0\n"
     ]
    }
   ],
   "source": [
    "def optimize_medium(bigg, type, exchange):\n",
    "    \"\"\"\n",
    "    This function does optimize based on a chemical in the certain biolog medium and add an exchange reaction if needed\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "    # when model has the exchange reaction, no need to add a reaction before gapfilling.    \n",
    "    with model:\n",
    "        medium = model.medium\n",
    "        medium[exchange] = 0.8\n",
    "        if type.startswith(\"C\"):\n",
    "            medium[\"EX_glc__D_e\"] = 0\n",
    "\n",
    "        elif type.startswith(\"N\"):\n",
    "            medium[\"EX_nh4_e\"] = 0\n",
    "\n",
    "        elif type.startswith(\"P\"):\n",
    "            medium[\"EX_pi_e\"] = 0\n",
    "        \n",
    "        elif type.startswith(\"S\"):\n",
    "            medium[\"EX_so4_e\"] = 0\n",
    "\n",
    "        print(f\"\\n Optimize {bigg}, {type} with: \")\n",
    "        model.medium = medium\n",
    "        print(model.medium)\n",
    "        model.solver = 'cplex'\n",
    "        model.objective = 'growth'\n",
    "        solution = model.optimize()\n",
    "        print(solution.objective_value)\n",
    "    return solution \n",
    "            \n",
    "            \n",
    "for ind in false_negatives.index:\n",
    "    optimize_medium(false_negatives['bigg'][ind], false_negatives['moa'][ind], false_negatives['exchange'][ind])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Gapfill Untestables with Sco model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_exchange(model, reference, metabolite_bigg, metabolite_name):\n",
    "    \"\"\"\n",
    "    This function aims to add an extra exchange reaction based on metabolite to the target model using reference model.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    metabolite_bigg: String\n",
    "    metabolite_name: String\n",
    "    model: Cobra Model\n",
    "    reference: Cobra Model\n",
    "    \"\"\"\n",
    "    try:\n",
    "        ex_rxn_id = f\"EX_{metabolite_bigg}_e\"\n",
    "        ex_met_id = f\"{metabolite_bigg}_e\"\n",
    "        cy_met_id = f\"{metabolite_bigg}_c\"\n",
    "        print(f\"Try to add {ex_rxn_id} reaction\")\n",
    "        if any(r.id == ex_rxn_id for r in model.reactions.query(ex_rxn_id)):\n",
    "            if model.reactions.get_by_id(ex_rxn_id).boundary == False:\n",
    "                model.add_boundary(model.metabolites.get_by_id(ex_met_id), type=\"exchange\")\n",
    "            return model, ex_rxn_id\n",
    "\n",
    "        else:\n",
    "            # Check if the reaction is already in reference model, \n",
    "            # if yes, just copy reaction to model,\n",
    "            # if not, add a new reaction.\n",
    "            \n",
    "                if reference.reactions.query(ex_rxn_id) != []:\n",
    "                    ex_rxn = reference.reactions.get_by_id(ex_rxn_id)\n",
    "                    model.add_reactions([ex_rxn.copy()])\n",
    "                    print (f\"Exchange: {ex_rxn_id} is added\")\n",
    "                \n",
    "                else:\n",
    "                    # Check if the metabolites is in target model in both cytosol and external\n",
    "                    # if yes, create new reaction based on reference model\n",
    "                    # if not, create new metabolite\n",
    "                    if any(m.id == cy_met_id for m in reference.metabolites.query(cy_met_id)):\n",
    "                        ex_rxn = Reaction(id=ex_rxn_id, name=f\"{metabolite_name}-exchange\")\n",
    "                        model.add_reactions([ex_rxn])\n",
    "                        met = reference.metabolites.get_by_id(cy_met_id).copy()\n",
    "                        met.id = ex_met_id\n",
    "                        met.compartment = 'e'\n",
    "                        ex_rxn.add_metabolites({met: -1})                        \n",
    "                        print(f\"Exchange: {ex_rxn_id} is added\")\n",
    "\n",
    "                    elif any(m.id == ex_met_id for m in reference.metabolites.query(ex_met_id)):\n",
    "                        ex_rxn = Reaction(id=ex_rxn_id, name=f\"{metabolite_name}-exchange\")\n",
    "                        model.add_reactions([ex_rxn])\n",
    "                        met = reference.metabolites.get_by_id(ex_met_id).copy()\n",
    "                        ex_rxn.add_metabolites({met: -1})                        \n",
    "                        print(f\"Exchange: {ex_rxn_id} is added\")\n",
    "                    else:\n",
    "                        print(f'{metabolite_bigg} is not in reference model\\n')\n",
    "                        return model, None\n",
    "                # Final check to see if the boundary is added\n",
    "                if model.reactions.get_by_id(ex_rxn_id).boundary == False:\n",
    "                    model.add_boundary(model.metabolites.get_by_id(met.id), type=\"exchange\")\n",
    "                return model, ex_rxn_id\n",
    "    except:\n",
    "        return model, None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Untestable conditions:\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bigg</th>\n",
       "      <th>name</th>\n",
       "      <th>moa</th>\n",
       "      <th>ignore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2hbut</td>\n",
       "      <td>2 Hydroxybutyrate C4H7O3</td>\n",
       "      <td>C-Source, carboxylic acid</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2obut</td>\n",
       "      <td>2-Oxobutanoate</td>\n",
       "      <td>C-Source, carboxylic acid</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2pg</td>\n",
       "      <td>D-Glycerate 2-phosphate</td>\n",
       "      <td>P-Source, organic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2pglyc</td>\n",
       "      <td>2-Phosphoglycolate</td>\n",
       "      <td>P-Source, organic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>35cgmp</td>\n",
       "      <td>3',5'-Cyclic GMP</td>\n",
       "      <td>P-Source, organic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>tym</td>\n",
       "      <td>Tyramine</td>\n",
       "      <td>N-Source, other</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>tyrp</td>\n",
       "      <td>Phosphotyrosine</td>\n",
       "      <td>P-Source, organic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>tyrp</td>\n",
       "      <td>Phosphotyrosine</td>\n",
       "      <td>P-Source, organic</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>urate</td>\n",
       "      <td>Urate C5H4N4O3</td>\n",
       "      <td>N-Source, other</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>val__D</td>\n",
       "      <td>D Valine</td>\n",
       "      <td>N-Source, amino acid</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      bigg                      name                        moa  ignore\n",
       "0    2hbut  2 Hydroxybutyrate C4H7O3  C-Source, carboxylic acid   False\n",
       "1    2obut            2-Oxobutanoate  C-Source, carboxylic acid   False\n",
       "2      2pg   D-Glycerate 2-phosphate          P-Source, organic   False\n",
       "3   2pglyc        2-Phosphoglycolate          P-Source, organic   False\n",
       "4   35cgmp          3',5'-Cyclic GMP          P-Source, organic   False\n",
       "..     ...                       ...                        ...     ...\n",
       "77     tym                  Tyramine            N-Source, other   False\n",
       "78    tyrp           Phosphotyrosine          P-Source, organic   False\n",
       "79    tyrp           Phosphotyrosine          P-Source, organic   False\n",
       "80   urate            Urate C5H4N4O3            N-Source, other   False\n",
       "81  val__D                  D Valine       N-Source, amino acid   False\n",
       "\n",
       "[82 rows x 4 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# All Untestable Growth Conditions ()\n",
    "print(f\"\\nUntestable conditions:\\n\")\n",
    "\n",
    "# Mapping of bigg and biolog model\n",
    "untestables = pd.read_csv(f\"{os.getcwd()}/../untestable_metabolites.csv\", converters={'Ignore': lambda x: True if x == 'TRUE' else False})\n",
    "\n",
    "# Join two tables together\n",
    "untestables = untestables.merge(agreed_bio_data[['activity', 'index', 'moa', 'bigg']], on='bigg', how='left')\n",
    "untestables['ignore'] = untestables['ignore'].replace({None: False, 'FALSE': True})\n",
    "untestables = untestables[[\"bigg\",\"name\", \"moa\", \"ignore\"]]\n",
    "# untestables.to_csv('untestables.csv')\n",
    "untestables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read LP format model from file /var/folders/wy/8d18n83n56j78d4j6zkz449m0000gn/T/tmpwb_3z1u2.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 1881 rows, 4514 columns, 19772 nonzeros\n",
      "---------------\n",
      "Gapfilling 2hbut, C-Source, carboxylic acid test:\n",
      "\n",
      "Try to add EX_2hbut_e reaction\n",
      "2hbut is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling 2obut, C-Source, carboxylic acid test:\n",
      "\n",
      "Try to add EX_2obut_e reaction\n",
      "Exchange: EX_2obut_e is added\n",
      "\n",
      "Gapfilling 2obut with: \n",
      "Gapfilling of the compound 2obut - C-Source, carboxylic acid - EX_2obut_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling 2pg, P-Source, organic test:\n",
      "\n",
      "Try to add EX_2pg_e reaction\n",
      "Exchange: EX_2pg_e is added\n",
      "\n",
      "Gapfilling 2pg with: \n",
      "Gapfilling of the compound 2pg - P-Source, organic - EX_2pg_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling 2pglyc, P-Source, organic test:\n",
      "\n",
      "Try to add EX_2pglyc_e reaction\n",
      "Exchange: EX_2pglyc_e is added\n",
      "\n",
      "Gapfilling 2pglyc with: \n",
      "Gapfilling of the compound 2pglyc - P-Source, organic - EX_2pglyc_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling 35cgmp, P-Source, organic test:\n",
      "\n",
      "Try to add EX_35cgmp_e reaction\n",
      "Exchange: EX_35cgmp_e is added\n",
      "\n",
      "Gapfilling 35cgmp with: \n",
      "Gapfilling of the compound 35cgmp - P-Source, organic - EX_35cgmp_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling 35cgmp, P-Source, organic test:\n",
      "\n",
      "Try to add EX_35cgmp_e reaction\n",
      "\n",
      "Gapfilling 35cgmp with: \n",
      "Gapfilling of the compound 35cgmp - P-Source, organic - EX_35cgmp_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling 3hoxpac, C-Source, carboxylic acid test:\n",
      "\n",
      "Try to add EX_3hoxpac_e reaction\n",
      "3hoxpac is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling 3pg, P-Source, organic test:\n",
      "\n",
      "Try to add EX_3pg_e reaction\n",
      "Exchange: EX_3pg_e is added\n",
      "\n",
      "Gapfilling 3pg with: \n",
      "Gapfilling of the compound 3pg - P-Source, organic - EX_3pg_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling 5aptn, C-Source, carboxylic acid test:\n",
      "\n",
      "Try to add EX_5aptn_e reaction\n",
      "Exchange: EX_5aptn_e is added\n",
      "\n",
      "Gapfilling 5aptn with: \n",
      "Gapfilling of the compound 5aptn - C-Source, carboxylic acid - EX_5aptn_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling 5aptn, N-Source, other test:\n",
      "\n",
      "Try to add EX_5aptn_e reaction\n",
      "\n",
      "Gapfilling 5aptn with: \n",
      "Gapfilling of the compound 5aptn - N-Source, other - EX_5aptn_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling 6ax, N-Source, other test:\n",
      "\n",
      "Try to add EX_6ax_e reaction\n",
      "6ax is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling 6pgc, P-Source, organic test:\n",
      "\n",
      "Try to add EX_6pgc_e reaction\n",
      "Exchange: EX_6pgc_e is added\n",
      "\n",
      "Gapfilling 6pgc with: \n",
      "Gapfilling of the compound 6pgc - P-Source, organic - EX_6pgc_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling acgal, C-Source, carbohydrate test:\n",
      "\n",
      "Try to add EX_acgal_e reaction\n",
      "acgal is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling acgal, N-Source, other test:\n",
      "\n",
      "Try to add EX_acgal_e reaction\n",
      "acgal is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling acgam, C-Source, carbohydrate test:\n",
      "\n",
      "Try to add EX_acgam_e reaction\n",
      "Exchange: EX_acgam_e is added\n",
      "\n",
      "Gapfilling acgam with: \n",
      "[[<Reaction ACGApts at 0x7f76b237b2e0>], [<Reaction ACGApts at 0x7f76b237b2e0>], [<Reaction ACGApts at 0x7f76b237b2e0>], [<Reaction ACGApts at 0x7f76b237b2e0>]]\n",
      "---------------\n",
      "Gapfilling acgam, N-Source, other test:\n",
      "\n",
      "Try to add EX_acgam_e reaction\n",
      "\n",
      "Gapfilling acgam with: \n",
      "[[<Reaction ACGApts at 0x7f76c43d3550>], [<Reaction ACGApts at 0x7f76c43d3550>], [<Reaction ACGApts at 0x7f76c43d3550>], [<Reaction ACGApts at 0x7f76c43d3550>]]\n",
      "---------------\n",
      "Gapfilling acmana, C-Source, carbohydrate test:\n",
      "\n",
      "Try to add EX_acmana_e reaction\n",
      "Exchange: EX_acmana_e is added\n",
      "\n",
      "Gapfilling acmana with: \n",
      "Gapfilling of the compound acmana - C-Source, carbohydrate - EX_acmana_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling acmana, N-Source, other test:\n",
      "\n",
      "Try to add EX_acmana_e reaction\n",
      "\n",
      "Gapfilling acmana with: \n",
      "Gapfilling of the compound acmana - N-Source, other - EX_acmana_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling ad, C-Source, amide test:\n",
      "\n",
      "Try to add EX_ad_e reaction\n",
      "ad is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling ad, N-Source, other test:\n",
      "\n",
      "Try to add EX_ad_e reaction\n",
      "ad is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling agm, N-Source, other test:\n",
      "\n",
      "Try to add EX_agm_e reaction\n",
      "Exchange: EX_agm_e is added\n",
      "\n",
      "Gapfilling agm with: \n",
      "[[<Reaction ARGAGMt7 at 0x7f76c6b96670>], [<Reaction ARGAGMt7 at 0x7f76c6b96670>], [<Reaction ARGAGMt7 at 0x7f76c6b96670>], [<Reaction ARGAGMt7 at 0x7f76c6b96670>]]\n",
      "---------------\n",
      "Gapfilling alltn, N-Source, other test:\n",
      "\n",
      "Try to add EX_alltn_e reaction\n",
      "Exchange: EX_alltn_e is added\n",
      "\n",
      "Gapfilling alltn with: \n",
      "[[<Reaction ALLTNt2r at 0x7f76909f62b0>], [<Reaction ALLTNt2r at 0x7f76909f62b0>], [<Reaction ALLTNt2r at 0x7f76909f62b0>], [<Reaction ALLTNt2r at 0x7f76909f62b0>]]\n",
      "---------------\n",
      "Gapfilling amp2p, P-Source, organic test:\n",
      "\n",
      "Try to add EX_amp2p_e reaction\n",
      "amp2p is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling arab__D, C-Source, carbohydrate test:\n",
      "\n",
      "Try to add EX_arab__D_e reaction\n",
      "Exchange: EX_arab__D_e is added\n",
      "\n",
      "Gapfilling arab__D with: \n",
      "Gapfilling of the compound arab__D - C-Source, carbohydrate - EX_arab__D_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling arbt, C-Source, carbohydrate test:\n",
      "\n",
      "Try to add EX_arbt_e reaction\n",
      "arbt is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling argp, P-Source, organic test:\n",
      "\n",
      "Try to add EX_argp_e reaction\n",
      "argp is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling asn__D, N-Source, amino acid test:\n",
      "\n",
      "Try to add EX_asn__D_e reaction\n",
      "asn__D is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling but, C-Source, carboxylic acid test:\n",
      "\n",
      "Try to add EX_but_e reaction\n",
      "but is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling camp, P-Source, organic test:\n",
      "\n",
      "Try to add EX_camp_e reaction\n",
      "Exchange: EX_camp_e is added\n",
      "\n",
      "Gapfilling camp with: \n",
      "Gapfilling of the compound camp - P-Source, organic - EX_camp_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling cbp, P-Source, organic test:\n",
      "\n",
      "Try to add EX_cbp_e reaction\n",
      "Exchange: EX_cbp_e is added\n",
      "\n",
      "Gapfilling cbp with: \n",
      "Gapfilling of the compound cbp - P-Source, organic - EX_cbp_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling CE0074, N-Source, other test:\n",
      "\n",
      "Try to add EX_CE0074_e reaction\n",
      "CE0074 is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling cholp, P-Source, organic test:\n",
      "\n",
      "Try to add EX_cholp_e reaction\n",
      "cholp is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling citr__L, N-Source, amino acid test:\n",
      "\n",
      "Try to add EX_citr__L_e reaction\n",
      "\n",
      "Gapfilling citr__L with: \n",
      "Gapfilling of the compound citr__L - N-Source, amino acid - EX_citr__L_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling ethamp, P-Source, organic test:\n",
      "\n",
      "Try to add EX_ethamp_e reaction\n",
      "ethamp is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling frmd, N-Source, other test:\n",
      "\n",
      "Try to add EX_frmd_e reaction\n",
      "Exchange: EX_frmd_e is added\n",
      "\n",
      "Gapfilling frmd with: \n",
      "Gapfilling of the compound frmd - N-Source, other - EX_frmd_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling g1p, C-Source, carbohydrate test:\n",
      "\n",
      "Try to add EX_g1p_e reaction\n",
      "Exchange: EX_g1p_e is added\n",
      "\n",
      "Gapfilling g1p with: \n",
      "Gapfilling of the compound g1p - C-Source, carbohydrate - EX_g1p_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling g1p, P-Source, organic test:\n",
      "\n",
      "Try to add EX_g1p_e reaction\n",
      "\n",
      "Gapfilling g1p with: \n",
      "Gapfilling of the compound g1p - P-Source, organic - EX_g1p_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling g6p, C-Source, carbohydrate test:\n",
      "\n",
      "Try to add EX_g6p_e reaction\n",
      "Exchange: EX_g6p_e is added\n",
      "\n",
      "Gapfilling g6p with: \n",
      "Gapfilling of the compound g6p - C-Source, carbohydrate - EX_g6p_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling g6p, P-Source, organic test:\n",
      "\n",
      "Try to add EX_g6p_e reaction\n",
      "\n",
      "Gapfilling g6p with: \n",
      "Gapfilling of the compound g6p - P-Source, organic - EX_g6p_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling gam, C-Source, carbohydrate test:\n",
      "\n",
      "Try to add EX_gam_e reaction\n",
      "Exchange: EX_gam_e is added\n",
      "\n",
      "Gapfilling gam with: \n",
      "[[<Reaction GAMpts at 0x7f76629462b0>], [<Reaction GAMpts at 0x7f76629462b0>], [<Reaction GAMpts at 0x7f76629462b0>], [<Reaction GAMpts at 0x7f76629462b0>]]\n",
      "---------------\n",
      "Gapfilling gam, N-Source, other test:\n",
      "\n",
      "Try to add EX_gam_e reaction\n",
      "\n",
      "Gapfilling gam with: \n",
      "[[<Reaction GAMpts at 0x7f76523d44c0>], [<Reaction GAMpts at 0x7f76523d44c0>], [<Reaction GAMpts at 0x7f76523d44c0>], [<Reaction GAMpts at 0x7f76523d44c0>]]\n",
      "---------------\n",
      "Gapfilling gam6p, P-Source, organic test:\n",
      "\n",
      "Try to add EX_gam6p_e reaction\n",
      "Exchange: EX_gam6p_e is added\n",
      "\n",
      "Gapfilling gam6p with: \n",
      "Gapfilling of the compound gam6p - P-Source, organic - EX_gam6p_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling glx, C-Source, carboxylic acid test:\n",
      "\n",
      "Try to add EX_glx_e reaction\n",
      "Exchange: EX_glx_e is added\n",
      "\n",
      "Gapfilling glx with: \n",
      "Gapfilling of the compound glx - C-Source, carboxylic acid - EX_glx_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling glyc2p, P-Source, organic test:\n",
      "\n",
      "Try to add EX_glyc2p_e reaction\n",
      "glyc2p is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling hxa, C-Source, carboxylic acid test:\n",
      "\n",
      "Try to add EX_hxa_e reaction\n",
      "Exchange: EX_hxa_e is added\n",
      "\n",
      "Gapfilling hxa with: \n",
      "Gapfilling of the compound hxa - C-Source, carboxylic acid - EX_hxa_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling mal__D, C-Source, carboxylic acid test:\n",
      "\n",
      "Try to add EX_mal__D_e reaction\n",
      "mal__D is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling mal__D, C-Source, carboxylic acid test:\n",
      "\n",
      "Try to add EX_mal__D_e reaction\n",
      "mal__D is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling malon, C-Source, carboxylic acid test:\n",
      "\n",
      "Try to add EX_malon_e reaction\n",
      "Exchange: EX_malon_e is added\n",
      "\n",
      "Gapfilling malon with: \n",
      "Gapfilling of the compound malon - C-Source, carboxylic acid - EX_malon_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling man1p, P-Source, organic test:\n",
      "\n",
      "Try to add EX_man1p_e reaction\n",
      "Exchange: EX_man1p_e is added\n",
      "\n",
      "Gapfilling man1p with: \n",
      "Gapfilling of the compound man1p - P-Source, organic - EX_man1p_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling man6p, P-Source, organic test:\n",
      "\n",
      "Try to add EX_man6p_e reaction\n",
      "Exchange: EX_man6p_e is added\n",
      "\n",
      "Gapfilling man6p with: \n",
      "Gapfilling of the compound man6p - P-Source, organic - EX_man6p_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling mbdg, C-Source, carbohydrate test:\n",
      "\n",
      "Try to add EX_mbdg_e reaction\n",
      "mbdg is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling minohp, P-Source, organic test:\n",
      "\n",
      "Try to add EX_minohp_e reaction\n",
      "minohp is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling pcreat, P-Source, organic test:\n",
      "\n",
      "Try to add EX_pcreat_e reaction\n",
      "pcreat is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling peamn, C-Source, amine test:\n",
      "\n",
      "Try to add EX_peamn_e reaction\n",
      "Exchange: EX_peamn_e is added\n",
      "\n",
      "Gapfilling peamn with: \n",
      "Gapfilling of the compound peamn - C-Source, amine - EX_peamn_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling peamn, N-Source, other test:\n",
      "\n",
      "Try to add EX_peamn_e reaction\n",
      "\n",
      "Gapfilling peamn with: \n",
      "Gapfilling of the compound peamn - N-Source, other - EX_peamn_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling pep, P-Source, organic test:\n",
      "\n",
      "Try to add EX_pep_e reaction\n",
      "Exchange: EX_pep_e is added\n",
      "\n",
      "Gapfilling pep with: \n",
      "Gapfilling of the compound pep - P-Source, organic - EX_pep_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling pser__D, P-Source, organic test:\n",
      "\n",
      "Try to add EX_pser__D_e reaction\n",
      "pser__D is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling pser__L, P-Source, organic test:\n",
      "\n",
      "Try to add EX_pser__L_e reaction\n",
      "Exchange: EX_pser__L_e is added\n",
      "\n",
      "Gapfilling pser__L with: \n",
      "Gapfilling of the compound pser__L - P-Source, organic - EX_pser__L_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling salcn, C-Source, carbohydrate test:\n",
      "\n",
      "Try to add EX_salcn_e reaction\n",
      "Exchange: EX_salcn_e is added\n",
      "\n",
      "Gapfilling salcn with: \n",
      "[[<Reaction SALCpts at 0x7f7625920520>], [<Reaction SALCpts at 0x7f7625920520>], [<Reaction SALCpts at 0x7f7625920520>], [<Reaction SALCpts at 0x7f7625920520>]]\n",
      "---------------\n",
      "Gapfilling thrp, P-Source, organic test:\n",
      "\n",
      "Try to add EX_thrp_e reaction\n",
      "Exchange: EX_thrp_e is added\n",
      "\n",
      "Gapfilling thrp with: \n",
      "Gapfilling of the compound thrp - P-Source, organic - EX_thrp_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling tmp, P-Source, inorganic test:\n",
      "\n",
      "Try to add EX_tmp_e reaction\n",
      "tmp is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling tym, C-Source, amine test:\n",
      "\n",
      "Try to add EX_tym_e reaction\n",
      "tym is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling tym, N-Source, other test:\n",
      "\n",
      "Try to add EX_tym_e reaction\n",
      "tym is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling tyrp, P-Source, organic test:\n",
      "\n",
      "Try to add EX_tyrp_e reaction\n",
      "tyrp is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling tyrp, P-Source, organic test:\n",
      "\n",
      "Try to add EX_tyrp_e reaction\n",
      "tyrp is not in reference model\n",
      "\n",
      "---------------\n",
      "Gapfilling urate, N-Source, other test:\n",
      "\n",
      "Try to add EX_urate_e reaction\n",
      "Exchange: EX_urate_e is added\n",
      "\n",
      "Gapfilling urate with: \n",
      "Gapfilling of the compound urate - N-Source, other - EX_urate_e is failed\n",
      "\n",
      "---------------\n",
      "Gapfilling val__D, N-Source, amino acid test:\n",
      "\n",
      "Try to add EX_val__D_e reaction\n",
      "val__D is not in reference model\n",
      "\n"
     ]
    }
   ],
   "source": [
    "_model = model.copy()\n",
    "with open('gapfill_reactions_sco_untestable.txt', 'w') as file:\n",
    "    for ind in untestables.index:\n",
    "        bigg = untestables['bigg'][ind]\n",
    "        type = untestables['moa'][ind]\n",
    "        name = untestables['name'][ind]\n",
    "\n",
    "        # Only test ones that needs to be added\n",
    "        if untestables['ignore'][ind] == False:\n",
    "            file.write(f\"Gapfilling {bigg}, {type} with:\\n\")\n",
    "            print(f\"---------------\\nGapfilling {bigg}, {type} test:\\n\")\n",
    "            # Try to add exchange reactions before gapfilling\n",
    "            _model, ex_rxn = add_exchange(_model, sco_model, bigg, name)\n",
    "            solution = None\n",
    "            if ex_rxn:\n",
    "                solution = gapfill_medium(_model, sco_model, bigg, type, ex_rxn)\n",
    "\n",
    "            # Write gapfilling results\n",
    "            if solution:\n",
    "                unique_solution = set([reaction for reaction_set in solution for reaction in reaction_set])\n",
    "                reactions_str = ', '.join(str(reaction.id) for reaction in unique_solution)\n",
    "                file.write(reactions_str + '\\n\\n')\n",
    "            else:\n",
    "                file.write('\\n\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ALLTNt2r', 'GAMpts', 'SALCpts', 'ARGAGMt7', 'ACGApts'}\n"
     ]
    }
   ],
   "source": [
    "# Initialize an empty list to store the extracted lists\n",
    "gapfill_lists = []\n",
    "\n",
    "# Open and read the file\n",
    "with open('gapfill_reactions_uni_untestable.txt', 'r') as file:\n",
    "    lines = file.readlines()\n",
    "\n",
    "# Process every second line and split it into a list\n",
    "for i in range(1, len(lines), 3):\n",
    "    reaction_list = [item.strip() for item in lines[i].strip().split(',')]\n",
    "    gapfill_lists.extend(reaction_list)\n",
    "\n",
    "merged_lists = set(gapfill_lists)\n",
    "merged_lists.discard('')\n",
    "\n",
    "# Print the extracted lists\n",
    "print(merged_lists)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "for reaction_id in merged_lists:\n",
    "    # Get the reaction from the sco_model model\n",
    "    reaction = sco_model.reactions.get_by_id(reaction_id)\n",
    "    \n",
    "    # Add a copy of the reaction to your model\n",
    "    model.add_reactions([reaction.copy()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# 5. Compare model before and after"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'' is not a valid SBML 'SId'.\n"
     ]
    }
   ],
   "source": [
    "# Compare two models by reaction numbers and metabolites.\n",
    "updated_model = read_sbml_model(f\"{os.getcwd()}/Salb-GEM-Sco-gapfill.xml\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>7fa688a85850</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1874</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>2233</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>1386</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*growth - 1.0*growth_reverse_5ff58</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>Cytoplasm, Extracellular</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model  at 0x7fa688a85850>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>7fa688a91640</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1886</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>2262</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>1405</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*growth - 1.0*growth_reverse_5ff58</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>Cytoplasm, Extracellular</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model  at 0x7fa688a91640>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "updated_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>Sco_GEM_v1_3_1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>7fa6a9781e20</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>2064</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>2599</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>1775</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>7</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*BIOMASS_SCO_tRNA - 1.0*BIOMASS_SCO_tRNA_reverse_d3620</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>Cytoplasm, Extracellular</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model Sco_GEM_v1_3_1 at 0x7fa6a9781e20>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sco_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.EX_co2_e.bounds= (0, 1000)\n",
    "\n",
    "cobra.io.write_sbml_model(model, 'Salb-GEM-Sco-gapfill.xml')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
